README.md
# js-framerate-optimizer

[![npm version](https://img.shields.io/npm/v/framerate-optimizer.svg?style=flat-square)](https://www.npmjs.com/package/framerate-optimizer)
[![build](https://img.shields.io/github/workflow/status/gkjohnson/js-framerate-optimizer/Node.js%20CI?style=flat-square&label=build)](https://github.com/gkjohnson/js-framerate-optimizer/actions)
[![lgtm code quality](https://img.shields.io/lgtm/grade/javascript/g/gkjohnson/js-framerate-optimizer.svg?style=flat-square&label=code-quality)](https://lgtm.com/projects/g/gkjohnson/js-framerate-optimizer/)

Library for tracking and iteratively improving framerate over time inspired by the [Babylon.SceneOptimizer](https://doc.babylonjs.com/how_to/how_to_use_sceneoptimizer).

The terms "increase work" and "decrease work" are synonymous with "improve quality" and "improve performance" respectively.

# Use

```js
import { Optimizer } from '.../Optimizer.js';

// wait for x milliseconds
function wait(ms) {

    const start = window.performance.now();
    while (window.performance.now < start + ms) {}

}

// optimize the work time until we reach the target framerate
const optimizer = new Optimizer();
optimizer.addOptimization(delta => {

    workTime += delta < 0 ? -1 : 1;
    return true;

});

// work loop
let workTime = 500;
(function loop(){

    optimizer.update();
    wait(workTime);
    requestAnimationFrame(loop);

})();

```

# How it Works

The optimizer tracks the amount of time spent between frames or between calls to `begin` and `end` and calculates the difference between the target amount of time to spend and the actual time spent on the last frame. After the specified amount of time has passed the average time spent is calculated and the framerate the amount of work is either increased or decreased depending on whether or not the time spent was above or below the target. The amount of work is adjusted by iteratively calling prioritized optimizations and sampling framerate until the target work time is met.

# API
## Optimizer

### .enabled

```js
enabled = true : Boolean
```

Getter and setter for enabling or disabling the optimizer. Elapsed time is reset on reenable.

### .completed

```js
completed : Boolean
```

Whether or not the optimizer has stopped iterating and sampling the framerate.


### .constructor

```js
constructor( options : Object )
```

All options can be accessed and modified on `Optimizer.options`.

#### options.targetMillis

The target milliseconds to hit in the enclosed code block. This cannot be _less_ than `16.66...` because browser caps the framerate to 60 frames per second. If this is less than that the optimizer will continually decrease quality to try to get performance up.

Defaults to `16.66...`.

#### options.targetFramerate

The target framerate to hit. This overrides `targetMillis` if it is set.

Defaults to `null`.

#### options.interval

How often to perform a check sample the average framerate to try to optimize in milliseconds.

Defaults to `500` or half a second.

#### options.maxFrameSamples

At most how many frames can run before an optimization should occur. This is useful when code may not run consistently but runs as needed, meaning that checking after a fixed amount of time might mean that an inconsistent or no actual iterations has occurred. `Interval` should probably be set to `Infinity` in this case.

Defaults to `Infinity`.

#### options.waitMillis

The amount of time to wait between sampling frames for a new optimization. This is useful when an optimization may cause the framerate to drop for a frame or two -- such as with allocating a lot of new memory or new webgl render targets.

Defaults to `0`;

#### options.maxWaitFrames

At most how many frames to wait for.

Defaults to `Infinity`;

#### options.margin

How far outside of the target framerate must be in order for an optimization to occurs.

Defaults to `0.05` or 5%.

#### options.continuallyRefine

By default the optimizer will stop trying to optimize and optimization the page once the target framerate is hit or no more optimizations can run and will never try to improve quality of the page again after the first iteration.

This option allows the optimizer to ping pong between improving performance and quality continually to keep a steady framerate. Note that this may not be a good option when using "expensive" optimizations that may stall the frame for a moment, like compiling shaders.

Defaults to `false`.

#### options.increaseWork

Whether the optimizer should ever try to increase the amount of work done at the cost of framerate. After the quality improvements are made the optimizer tries to improve the framerate until the target framerate is met.

Defaults to `false`.

### .dispose

```js
dispose() : void
```

Removes window events that the optimizer listens for, including window `"blur"` and `"focus"`. It is expected that the optimizer is no longer used after this.

### .addOptimization

```js
addOptimization( optimization : Optimization, priority = 0 : Number ) : void
```

Takes an instance of a `Optimization` (described below) and a priority by which to call the optimization. Optimizations are optimized iteratively until each priority level has been completely optimized then it will move to the next one. When improving quality the lowest priority level is started with to improve quality. The highest is started with when improving performance.

Expensive but unnecessary work should be optimized at a high priority level. Critical functionality should be optimized at a low priority level.

### .begin
### .end

```js
begin() : void
end() : void
```

`begin` an `end` are used to encapsulate the code block to be optimized. Any optimizations being used should affect the performance of the code within that code block.

```js
optimizer.begin();

// ... code to optimize ...

optimizer.end();
```

### .update

```js
update() : void
```

`update` takes the place of both `begin()` and `end()` and should be called once per frame to optimize the full frame time.

```js

function loop() {

    optimizer.update();

    // ... code to optimize

    requesAnimationFrame(loop);

}
```

### .addSample

```js
addSample( time : Number ) : void
```

Call instead of `begin`, `end`, or `update` if you'd like to use a separate time-measuring mechanism that cannot be achieved with the aforementioned function, such as when measuring GPU work time with the [WebGL disjoint_timer_query](https://www.khronos.org/registry/webgl/extensions/EXT_disjoint_timer_query_webgl2/).

### .restart

```js
restart() : void
```

Restarts optimization from the beginning.

This should be called if something on the page changed
icantly which might give more room for improvement, such as significantly resizing the browser window with a WebGL app.

## Optimization

An optimization base class for creating a reusable optimization to add to the `Optimizer`.

### .optimize

```js
optimize( delta : Number, optimizer : Optimizer )
```

The optimize function is called whenever an optimization should take place, either to improve performance or quality. The `delta` argument is the amount of milliseconds difference between the target framerate and the current framerate. A negative value means that less work should be done to hit the target.

The optimize function _must_ return `true` if the setting was optimized and 'false' if no optimization could occur, ie the setting being optimized could not be turned down any further or is at the lowest acceptable setting.

The second argument `optimizer` is the optimizer that is running the optimization.

## SimpleOptimization

A extension of `Optimization` that abstracts some of the basic functionality needed to implement it.

### .canIncreaseWork / .canDecreaseWork

```js
canIncreaseWork() : Boolean
canDecreaseWork() : Boolean
```

Functions that should return true if the optimizer _can_ increase or decrease work. If they return true the associated optimization function will be called.

### .increaseWork / .decreaseWork

```js
increaesWork( delta : Number ) : void
decreaseWork( delta : Number ) : void
```

Functions to increase or decrease work and optimize the frame if needed. These are called if the associated `canIncreaseWork` or `canDecreaseWork` return true.



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [0.2.0] - 2021-04-17
### Added
- `addSample` function.

### Changed
- Add `"type": "module"` to package.json.

## [0.1.2] - 2018-11-04
### Fix
- SimpleOptimization calling calling `increaseWork` incorrectly.

## [0.1.1] - 2018-11-02
### Fix
- Correctly name option "maxWaitFrames" instead of "maxwaitFrames"

## [0.1.0] - 2018-11-02
### Changed
- Moved module to src folder
- Move all options onto an options sub-object

### Added
- UMD variant to umd folder
- Add "waitMillis" and "maxWaitFrames" options

## [0.0.2] - 2018-08-26
### Fixed
- Properly remove events from window on dispose.

## [0.0.1] - 2018-08-10
- Initial release


umd/Optimizer.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
    typeof define === 'function' && define.amd ? define(['exports'], factory) :
    (factory((global.Optimizer = {})));
}(this, (function (exports) { 'use strict';

    class Optimization {

        optimize(delta, optimizer) {

            return true;

        }

    }

    class SimpleOptimization extends Optimization {

        canIncreaseWork() { return true; }
        increaseWork(delta) { }

        canDecreaseWork() { return true; }
        decreaseWork(delta) { }

        optimize(delta) {

            if (delta < 0) {

                if (this.canDecreaseWork()) {

                    this.decreaseWork(delta);
                    return true;

                }

            } else {

                if (this.canIncreaseWork()) {

                    this.increaseWork(delta);
                    return true;

                }

            }

            return false;

        }

    }

    class Optimizer {

        get enabled() {

            return this._enabled;

        }

        set enabled(val) {

            if (this._enabled !== val) {

                this.resetCheck();

            }

            this._enabled = val;

        }

        constructor(options) {

            options = Object.assign({

                // target milliseconds to hit in the code enclosed by
                // the optimizer
                targetMillis: 1000 / 60,
                targetFramerate: null,

                // how often to check performance
                interval: 500,
                maxFrameSamples: Infinity,

                // how long to wait between capturing frames
                waitMillis: 0,
                maxWaitFrames: Infinity,

                // how far outside the current framerate must be outside
                // the target to optimize
                margin: 0.05,

                // continue to improve quality and then performance over time
                // instead of just stopping after a single failed improvement.
                continuallyRefine: false,

                // whether not we're currently increasing the amount of work
                // done per frame (and thereby decreasing framerate)
                increaseWork: false,

            }, options);

            // convert the specified framerate option to millis
            if (options.targetFramerate > 0) {

                options.targetMillis = 1000 / options.targetFramerate;
                delete options.targetFramerate;

            }

            Object.defineProperty(options, 'targetFramerate', {
                get() { return 1000 / this.targetMillis; },
                set(v) { this.targetMillis = 1000 / v; },
            });

            this.options = options;

            this._enabled = true;
            this.completed = false;
            this._increasingWork = this.options.increaseWork;

            // the prioritized optimizations -- int : array
            // It would be best if this were sorted linked list so
            // large gaps don't cause unnecessary iteration
            this.optimizations = {};
            this.minPriority = Infinity;
            this.maxPriority = -Infinity;

            // Tracking the time between optimizations
            this.waitedFrames = this.options.maxWaitFrames;
            this.waitedMillis = this.options.waitMillis;
            this.elapsedFrames = 0;
            this.elapsedTime = 0;
            this.beginTime = -1;

            // The next optimization to try
            this.currPriority = null;
            this.currOptimization = 0;

            this._windowFocused = true;
            this._windowBlurFunc = () => this._windowFocused = false;
            this._windowFocusFunc = () => {

                this._windowFocused = true;
                this.resetCheck();

            };
            window.addEventListener('blur', this._windowBlurFunc);
            window.addEventListener('focus', this._windowFocusFunc);

        }

        dispose() {

            window.removeEventListener('blur', this._windowBlurFunc);
            window.removeEventListener('focus', this._windowFocusFunc);

        }

        /* Public API */
        // restarts the optimization process by first improving quality then
        // performance
        restart() {

            this.resetCheck();

            this._increasingWork = this.options.increaseWork;
            this.currPriority = null;
            this.currOptimization = 0;
            this.completed = false;

        }

        addSample(sampleTime) {

            // if we're not active for any reason, continue
            if (!this._enabled || !this._windowFocused || this.completed) return;

            // wait the required number of frames between calls
            if (this.waitedFrames !== 0 && this.waitedMillis !== 0) {
                this.waitedMillis -= sampleTime;
                this.waitedFrames--;

                this.waitedFrames = Math.max(this.waitedFrames, 0);
                this.waitedMillis = Math.max(this.waitedMillis, 0);
                return;
            }

            // increment the time and frames run
            this.elapsedTime += sampleTime;
            this.elapsedFrames++;

            // if we've waited for an appropriate amount of time
            if (this.elapsedTime >= this.options.interval || this.elapsedFrames >= this.options.maxFrameSamples) {

                // average time per frame and the differences
                const frameTime = this.elapsedTime / this.elapsedFrames;
                const delta = this.options.targetMillis - frameTime;
                const ratio = delta / this.options.targetMillis;
                const isOutsideMargin = Math.abs(ratio) > this.options.margin;
                const needsImproving = delta < 0 && isOutsideMargin;

                if (this._increasingWork) {

                    if (this.currPriority === null) {

                        this.currPriority = this.minPriority;

                    }

                    // If our frame time is higher than we want, then
                    // start trying to improve it.
                    if (needsImproving) {

                        this._increasingWork = false;
                        this.currPriority = this.maxPriority;
                        this.currOptimization = 0;

                    } else {

                        // delta will always be ~0 when targeting 60 fps because the
                        // browser runs at a fixed framerate
                        this.iterate(Math.max(delta, 1));

                    }

                }

                // Try to improve the frame time
                if (!this._increasingWork) {

                    if (this.currPriority === null) {

                        this.currPriority = this.maxPriority;

                    }

                    let didOptimize = false;

                    if (needsImproving) {

                        didOptimize = this.iterate(delta);

                    }

                    if (!didOptimize) {

                        if (this.options.continuallyRefine) {

                            this._increasingWork = true;

                        } else {

                            this.completed = true;

                        }

                    }

                }

                this.elapsedFrames = 0;
                this.elapsedTime = 0;
                this.waitedFrames = this.options.maxWaitFrames;
                this.waitedMillis = this.options.waitMillis;

            }

        }

        // begin the code block to optimize
        begin() {

            this.beginTime = window.performance.now();

        }

        // end the code block to optimize
        end() {

            // If end is called before begin then skip this iteration
            if (this.beginTime === -1) return;

            const timeFromBegin = window.performance.now() - this.beginTime;
            this.addSample(timeFromBegin);

        }

        // A single function to use _instead_ of "begin" and "end". The function
        // should be called once per frame to optimize on the full frame time
        update() {

            this.end();
            this.begin();

        }

        // add a optimization function at the given priority
        addOptimization(optimization, priority = 0) {

            if (typeof optimization === 'function') {

                const optimizationFunc = optimization;
                optimization = new Optimization();
                optimization.optimize = optimizationFunc;

            }

            priority = parseInt(priority) || 0;
            this.optimizations[priority] = this.optimizations[priority] || [];
            this.optimizations[priority].push(optimization);

            this.minPriority = Math.min(this.minPriority, priority);
            this.maxPriority = Math.max(this.maxPriority, priority);

        }

        /* Private Functions */
        // Iterates over the optimizations based on the delta. Improving quality if delta > 0
        // and performance if delta < 0
        iterate(delta) {

            let done = false;
            while (this.currPriority <= this.maxPriority && this.currPriority >= this.minPriority) {

                // search for a optimization we can perform to improve performance
                // if we get through all optimizations without an improvement then
                // move on to the next priority level.
                const optimizations = this.optimizations[this.currPriority];
                if (optimizations) {

                    for (let i = 0; !done && i < optimizations.length; i++) {

                        done = optimizations[this.currOptimization].optimize(delta, this);

                        if (typeof done !== 'boolean') {

                            done = !!done;
                            console.warn('Optimizer: Optimization function not returning a boolean value.');

                        }

                        this.currOptimization = (this.currOptimization + 1) % optimizations.length;

                    }

                }

                if (done) {

                    break;

                } else {

                    // Lower priority numbers are more important
                    this.currPriority += delta > 0 ? 1 : -1;
                    this.currOptimization = 0;

                }

            }

            return done;

        }

        // resets the current frame check
        resetCheck() {

            this.elapsedFrames = 0;
            this.elapsedTime = 0;
            this.waitedFrames = this.options.maxWaitFrames;
            this.waitedMillis = this.options.waitMillis;
            this.beginTime = -1;

        }

    }

    exports.Optimization = Optimization;
    exports.SimpleOptimization = SimpleOptimization;
    exports.Optimizer = Optimizer;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
